xdsl/ir/core.py

```diff
@@ -373,41 +410,125 @@ class MyEnumAttribute(EnumAttribute[MyEnum], SpacedOpaqueSyntaxAttribute):
-        reasonable, until more complex use cases appear. It then stores the Enum type
-        used by the subclass to use in parsing/printing.
+        _check_enum_constraints(cls)
 
-        The constraints are:
+    def print_parameter(self, printer: Printer) -> None:
+        printer.print(self.data.value)
 
-        - Only direct, specialized inheritance is allowed. That is, using a subclass
-        of EnumAttribute as a base class is *not supported*.
-          This simplifies type-hacking code and I don't see it being too restrictive
-          anytime soon.
-        - The StrEnum values must all be parsable as identifiers. This is to keep the
-        parsing code simple and efficient. This restriction is easier to lift, but I
-        haven't yet met an example use case where it matters, so I'm keeping it simple.
-        """
-        orig_bases = getattr(cls, "__orig_bases__")
-        enumattr = next(b for b in orig_bases if get_origin(b) is EnumAttribute)
-        enum_type = get_args(enumattr)[0]
-        if isinstance(enum_type, TypeVar):
-            raise TypeError("Only direct inheritance from EnumAttribute is allowed.")
+    @classmethod
+    def parse_parameter(cls, parser: AttrParser) -> EnumType:
+        return cast(EnumType, parser.parse_str_enum(cls.enum_type))
 
-        for v in enum_type:
-            if lexer.Lexer.bare_identifier_suffix_regex.fullmatch(v) is None:
-                raise ValueError(
-                    "All StrEnum values of an EnumAttribute must be parsable as an identifer."
+
+class BitEnumAttribute(Generic[EnumType], Data[tuple[EnumType, ...]]):
+    """
+    Core helper for BitEnumAttributes. Takes a StrEnum type parameter, and
+    defines parsing/printing automatically from its values.
+
+    Additionally, two values can be given to designate all/none bits being set.
+
+    example:
+    ```python
+    class MyBitEnum(StrEnum):
+        First = auto()
+        Second = auto()
+
+    class MyBitEnumAttribute(BitEnumAttribute[MyBitEnum]):
+        name = "example.my_bit_enum"
+        none_value = "none"
+        all_value = "all"
+
+    """
+
+    enum_type: ClassVar[type[StrEnum]]
+    none_value: ClassVar[str | None] = None
+    all_value: ClassVar[str | None] = None
+
+    def __init__(self, flags: None | Sequence[EnumType] | str) -> None:
+        flags_: set[StrEnum]
+        match flags:
+            case self.none_value | None:
+                flags_ = set()
+            case self.all_value:
+                flags_ = set(self.enum_type)
+            case other if isinstance(other, str):
+                raise TypeError(
+                    f"expected string parameter to be one of {self.none_value} or {self.all_value}, got {other}"
                 )
+            case other:
+                assert not isinstance(other, str)
+                flags_ = set(other)
 
-        cls.enum_type = enum_type
+        super().__init__(tuple(cast(set[EnumType], flags_)))
```

Comment:
Hmm, the branches do set it to StrEnum, set[StrEnum] does not equal set[EnumType]
```
Expression of type "set[StrEnum] is imcompatible with return type "set[EnumType]"
```
Even though `EnumType = TypeVar('EnumType', bound=StrEnum)`

This is because cls.enumtype is `StrEnum`. I cannot hovewer make this `EnumType`, as it is illegal to have a typevar as a classvariable